bpf/lib/fib.h

```diff
@@ -93,94 +106,172 @@ fib_redirect(struct __ctx_buff *ctx, const bool needs_l2_check,
+				else
+					return redirect_neigh(*oif, NULL, 0, 0);
+			} else {
+				union macaddr *dmac, smac = NATIVE_DEV_MAC_BY_IFINDEX(*oif);
+
+				/* The neigh_record_ip{4,6} locations are mainly from
+				* inbound client traffic on the load-balancer where we
+				* know that replies need to go back to them.
+				*/
+				dmac = fib_params->l.family == AF_INET ?
+				neigh_lookup_ip4(&fib_params->l.ipv4_dst) :
+				neigh_lookup_ip6((void *)&fib_params->l.ipv6_dst);
+				if (!dmac) {
+					*fib_ret = BPF_FIB_MAP_NO_NEIGH;
+					return DROP_NO_FIB;
+				}
+				if (eth_store_daddr_aligned(ctx, dmac->addr, 0) < 0)
+					return DROP_WRITE_ERROR;
+				if (eth_store_saddr_aligned(ctx, smac.addr, 0) < 0)
+					return DROP_WRITE_ERROR;
+			}
+	};
 out_send:
 	return ctx_redirect(ctx, *oif, 0);
 }
 
-#ifdef ENABLE_IPV6
 static __always_inline int
-fib_redirect_v6(struct __ctx_buff *ctx, int l3_off,
-		struct ipv6hdr *ip6, const bool needs_l2_check,
-		__s8 *fib_err, int iif, int *oif)
+fib_redirect(struct __ctx_buff *ctx, const bool needs_l2_check,
+	     struct bpf_fib_lookup_padded *fib_params __maybe_unused,
+	     __s8 *fib_err __maybe_unused, int *oif)
 {
-	struct bpf_fib_lookup_padded fib_params = {
-		.l = {
-			.family		= AF_INET6,
-			.ifindex	= iif,
-		},
-	};
+#ifndef ENABLE_SKIP_FIB
 	int ret;
+	ret = fib_lookup(ctx, &fib_params->l, sizeof(fib_params->l), 0);
+	*fib_err = (__s8)ret;
+	return fib_do_redirect(ctx, needs_l2_check, fib_params, fib_err, oif);
+#else
+	__s8 skip_fib = BPF_FIB_LKUP_RET_NO_NEIGH;
+	*oif = DIRECT_ROUTING_DEV_IFINDEX;
+	return fib_do_redirect(ctx, needs_l2_check, NULL, &skip_fib, oif);
+#endif
+}
+
+#ifdef ENABLE_IPV6
+/* fib_lookup_v6 will take a ctx along with a ipv6 header and parameters, perform
+ * a fib lookup with the src and dest addresses in the ipv6 header and return
+ * the code.
+ *
+ * after the function returns `fib_params` will have the results of the fib lookup
+ * if successful.
+ */
+static __always_inline int
+fib_lookup_v6(struct __ctx_buff *ctx, struct bpf_fib_lookup_padded *fib_params, struct ipv6hdr *ip6, int flags)
+{
+	fib_params->l.family	= AF_INET6;
+	fib_params->l.ifindex	= ctx_get_ifindex(ctx);
 
-	ipv6_addr_copy((union v6addr *)&fib_params.l.ipv6_src,
+	ipv6_addr_copy((union v6addr *)&fib_params->l.ipv6_src,
 		       (union v6addr *)&ip6->saddr);
-	ipv6_addr_copy((union v6addr *)&fib_params.l.ipv6_dst,
+	ipv6_addr_copy((union v6addr *)&fib_params->l.ipv6_dst,
 		       (union v6addr *)&ip6->daddr);
 
+	return fib_lookup(ctx, &fib_params->l, sizeof(fib_params->l), flags);
+};
+
+static __always_inline int
+fib_redirect_v6(struct __ctx_buff *ctx, int l3_off,
+		struct ipv6hdr *ip6 __maybe_unused, const bool needs_l2_check,
+		__s8 *fib_err __maybe_unused, int iif, int *oif)
+{
+	struct bpf_fib_lookup_padded fib_params __maybe_unused = {0};
+	void *data_end;
+	void *data;
+	__s8 skip_fib __maybe_unused = BPF_FIB_LKUP_RET_NO_NEIGH;
+	int ret;
+
+	/* TODO: Fix this*/
+	(void)iif;
```

Comment:
This is here since all invocation of `fib_redirect_*` get the iif from the ctx, which occurs now in fib_lookup.

Moving forward, if there's no reason to pass `iif` into the function, it can be removed from the signature, but deferring that to a later commit to update call sights.